File: Python/Tree/M-450. Delete Node in a BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        def dfs(node: Optional[TreeNode], key: int):
            if node is None:
                return None
            if node.val == key:
                if node.left is None:
                    return node.right
                if node.right is None:
                    return node.left
                cur = node.right
                while cur.left:
                    cur = cur.left
                cur.left = node.left
                return node.right
            if key < node.val:
                node.left = dfs(node.left, key)
            else:
                node.right = dfs(node.right, key)
            return node

        return dfs(root, key)

File: Tree/M-450. Delete Node in a BST/M-450. Delete Node in a BST.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function deleteNode(root: TreeNode | null, key: number): TreeNode | null {
  const dfs = (node: TreeNode | null, key: number) => {
    if (!node) return null;
    if (node.val === key) {
      if (!node.left) return node.right;
      if (!node.right) return node.left;
      let cur = node.right;
      while (cur.left) cur = cur.left;
      cur.left = node.left;
      return node.right;
    }
    if (key < node.val) node.left = dfs(node.left, key);
    else node.right = dfs(node.right, key);
    return node;
  };
  return dfs(root, key);
}
